Numark Mixstream Pro & Pro GO mapping

audministrator opened the pull request:
This PR holds the first **Numark Mixstream Pro & Pro GO** controller mapping.

**Refactoring and fixes:** Improved jogWheel, touchWheel, and Loop functionality, with better handling of functions.

Versus last version published https://mixxx.discourse.group/t/numark-mixstream-pro-mapping/24858/33?u=ptrex

**New features:** Added settings controls, brake/soft start options (ready for 2.6+), revamp EQ knobs, selector knob, pitch bend buttons alternative mapping, and updated beat sync to use sync_enabled.
**Code optimization:** reviewed code for un-used functions and variable scoping
**Mapping upgrades:** Lots of changes versus last published **version v12** here 
https://mixxx.discourse.group/t/numark-mixstream-pro-mapping/24858/33?u=ptrex

audministrator committed:
Add files via upload

audministrator committed:
Code correction

Replaced == by ===

audministrator reviewed (commented):
Code Style correction

ronso0 commented on res/controllers/Mixstream Pro.midi.xml:
```diff
@@ -0,0 +1,841 @@
+            <control>
+                <group>[Skin]</group>
+                <key>show_microphone</key>
+                <description>MIDI Learned from 17 messages.</description>
```
Please remove all those "MIDI learned ..." messages, they are not helpful in included mappings. Instead, leave it empty (remove `<description>` entirely) or enter something meaningful, for example "Mic" or "Mixer" that allows sorting controls by category in the UI.
```suggestion
                <description>Mic</description>
```

audministrator replied:
Done

ronso0 commented on res/controllers/Mixstream Pro.midi.xml:
```diff
@@ -0,0 +1,841 @@
+            <control>
+                <group>[Channel1]</group>
+                <key>MixstreamPro.EffectToggleSwitch</key>
+                <description>MIDI Deck1 Effect Toggle Switch</description>
```
Same here, MIDI is obsolete.

audministrator replied:
Done

audministrator commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+ 
+MixstreamPro.shutdown = function() {
+     
+    midi.sendShortMsg(0x90, 0x75, 0x00);
```
This is by Numark desgn.

1 midi call shuts down all LEDs 🙂
This mimics the orignal Numark Pro behavior..

ronso0 commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+
+            MixstreamPro.WIFI = false
+        } else
+        if (MixstreamPro.WIFI === false) {
```
```suggestion
        } else {
```

ronso0 commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+}
+
+ 
+MixstreamPro.WIFI = true
```
Can you explain what this does?

ronso0 commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+MixstreamPro.previousJogValue2 = 0
+
+MixstreamPro.WheelTouch = function (channel, control, value, status, group) {
+    let deckNumber = script.deckFromGroup(group)
```
below the indentation is off

ronso0 commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+MixstreamPro.JogLSB_1 = function(channel, control, value, status, group) {
+    
+    return 
+};
```
why ignore the LSB of the jogwheels?

Besides, why are there also two identical jogwheel functions?

ronso0 commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+MixstreamPro.hotcuevalue1 = 0
+MixstreamPro.LEDblink4 = true
+ 
+MixstreamPro.track_loaded1 = function(channel, control, value, status, group) {
```
Is this a callback function?
If no, how is it called?

Same here, refactor to one function and use `group`

ronso0 commented:
I left some more comments.
I strongly recommend to
1. work locally and make use of pre-commit
2. remove all the duplicate per-deck functions and make use of `group`

audministrator commented:
Hi @ronso0 

Maybe I just should stick to DJ instead of trying to code 🤔
Best remove the pull request, I will leave it someone else for now to get published...

Anyhow thanks for the review !!

ronso0 commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+        }
+    }
+    if (value === 0) { return }
+}
```
So IIUC this is meant to be a toggle for Aux1, play, right?
And assign Aux to the opposite side of the playing deck.
(btw what should happen if both decks are playing?)

Thing with the tracking bool `MixstreamPro.WIFI` is that it doesn't consider the actual engine state, like if you toggle Aux1 Play in the GUI, next press of the WIFI button can be a no-op if `[Auxiliary1], master` is already `1` (besides the correct control is [`[AuxiliaryN],main_mix`](https://manual.mixxx.org/2.5/en/chapters/appendix/mixxx_controls.html#control-[MicrophoneN]-main_mix)).
So IMO you can drop `MixstreamPro.WIFI`and do something like this
```
MixstreamPro.playAux1 = function(channel, control, value, status, group) {
    if (!value) {
        return;
    }

    if (engine.getValue("[Auxiliary1]", "main_mix")) {
        // Aux is off, turn it on
        engine.setValue("[Auxiliary1]", "main_mix", 1)
         
        if (engine.getValue("[Channel1]", "play") === 1) {
            engine.setValue("[Auxiliary1]", "orientation", 2)
        } else if (engine.getValue("[Channel2]", "play") === 1) {
            engine.setValue("[Auxiliary1]", "orientation", 0)
        }
    } else {
        // Aux is on, turn it off
        engine.setValue("[Auxiliary1]", "main_mix", 0)
    }
}
```
Note that I dropped `engine.setValue("[Auxiliary1]", "pregain", 0)`. The pregain is certainly a user setting and depends on a) the output level of the attached device and b) on desired level in the main mix.

audministrator commented on res/controllers/Mixstream Pro.midi.xml:
```diff
@@ -0,0 +1,841 @@
+<?xml version='1.0' encoding='utf-8'?>
+<MixxxControllerPreset mixxxVersion="" schemaVersion="1">
+    <info>
+        <name>Mixstream Pro v17</name>
```
Done

audministrator committed:
changed Controller name in XML

audministrator committed:
Update all XML Descriptions

audministrator commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+            playStatus === 1 ? engine.setValue(group, "play", 0) : engine.setValue(group, "play", 1);
+        }
+    }
+} 
```
_These can be just one playButton function, then simply use the group parameter_ 

I planned this in a next version update to condense the number of lines of code.
But if you prefer this in v1.0 I can do it but it will take a few weeks unfortunately.

audministrator commented:
Hmmm, 
I am on Windows 10 & 11 and have nothing installed (no python / node / etc) 
And don't intend to install any kind of developement env. so far if not needed.

audministrator committed:
Update Numark-MixstreamPro-scripts.js

Updated code errors : 
!=== replaced by !==
Updated let deckNumber by const deckNumber

ronso0 commented on res/controllers/Numark-MixstreamPro-scripts.js:
```diff
@@ -0,0 +1,1604 @@
+        midi.sendShortMsg(0x94, 1, 0x01);
+        midi.sendShortMsg(0x94, 2, 0x01);
+        midi.sendShortMsg(0x94, 3, 0x01);
+    }, 1);
```
What does the value `0x01` mean? Dim LED or bright LED?

audministrator replied:
0x01 = Dim LED

ronso0 replied:
Okay, so they are always dim, and when turned 'on' via MIDI they are bright?

Nevertheless, I recommend shortening the timeouts, or turn all LEDs on at once.
Thing is, there is no point in mimicing the original (standalone) behavior when it doesn't match Mixxx. Like the GUI may already be loaded and the controller appears to be still in "startup mode" (while it's actually ready way earlier)

audministrator replied:
Never experienced this behavior as you describe ?
But if it makes you happy, I can shorten it as down not problem.

I do this after I have done the the other major open topics, that you been commenting the last couple of days.
There are still quite a lot of things to do, but that's for next year when I have more time available....

ronso0 replied:
Sure, whenever it suits.

And this is not about me, it's about UX (user experience). Ideally all Mixxx interfaces (GUI, controller, controller screens)  should signal "ready!" when Mixxx is ready.

audministrator committed:
Code correction

Fixed error on line 1434 
Replaced ==== by ===

audministrator committed:
Code correction

Pre-pend un-used function parameters

audministrator reviewed (commented):
Code Correction and Code Format corrections
Prepend un-used function parameters as requested.

PS : Changes have not yet been validated on the controller.

All other changes & remarks are for later (after New Year probably)

audministrator committed:
Remove duplicate LoadSelectedTrack 

Removed duplicate LoadSelectedTrack key in Channel2 control.

audministrator committed:
Delete res/controllers/Numark-MixstreamPro-scripts.js

Delete res/controllers/Numark-MixstreamPro-scripts.js

audministrator committed:
Delete res/controllers/Mixstream Pro.midi.xml

Delete res/controllers/Mixstream Pro.midi.xml

audministrator committed:
Add files via upload

audministrator reviewed (commented):
Loaded 2 new files to be reiewed.
Code has been reduced by 50%.

ronso0 commented:
Superseded by #15725
